lib/async_hooks.js

```diff
@@ -207,6 +208,128 @@ class AsyncResource {
   }
 }
 
+// AsyncContext part
+const idResourcePairing = new Map();
+const contexts = new Set();
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // save new resource
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResourceWeak = idResourcePairing.get(triggerAsyncId);
+    contexts.forEach((context) => {
+      context._propagate(resource, triggerResourceWeak && triggerResourceWeak.get(), triggerAsyncId);
```

Comment:
Seems the context propagation here follows the trigger resource. Another option to choose here would be the execution resource. see https://github.com/nodejs/node/pull/21313#issuecomment-414638795 for some discussion in this. In short both variants my be seen as the "correct" one.

The big question here is if we should hardwire this decision in node core or  if this should be configurable.

Reply:
I did some more investigations in this. Following sample seems to loose context accross `await`:
```
async function testAwait() {
  async function foo() {}

  const asyncContext = new AsyncContext();
  asyncContext.enter().set("key", "value");
  await foo();
  assert.notStrictEqual(asyncContext.getStore(), undefined);  // asserts
  assert.strictEqual(asyncContext.getStore().get("key"), "value");
}
```
If the execution path is used instead the trigger path this works.

I looked into some implementations in the wild and found different solutions:
* always use execution path
* use execution path is present, fallback to trigger path
* use execution path for promises and trigger path for the rest
* use async_hooks only for promises

Reply:
WOW @Flarna ! I will dig into that. Thanks for the heads up!

Reply:
I think a copy on write should be not that expensive assuming that there are by far more async operations then changes of the value. Meanwhile I think the `enter()` API is effective the same then setting a new value to `AsyncLocal.value`. It depends also on the implementation. I played around to create an `AsyncLocal` class (see https://gist.github.com/Flarna/690fbb2c23902ec84075f0acfa4cc19c). It's not complete and I'm not yet happy with it as it is.

In general my assumption is that the number of independent `AsyncContext`/`AsyncLocal` instances should be quite low in real applications. if a lot different values need to be transferred this way it's always possible to put an arbitrary complex object on a `AsyncContext`.